batch/batch/worker/worker.py

```diff
@@ -2135,14 +2145,15 @@ async def cleanup_old_images(self):
                 for image_id in list(self.image_data.keys()):
                     now = time_msecs()
                     image_data = self.image_data[image_id]
-                    if image_data.ref_count == 0 and (now - image_data.last_accessed) > 10 * 60 * 1000:
-                        assert image_id != BATCH_WORKER_IMAGE_ID
-                        log.info(f'Found an unused image with ID {image_id}')
-                        await check_shell(f'docker rmi -f {image_id}')
-                        image_path = f'/host/rootfs/{image_id}'
-                        await blocking_to_async(self.pool, shutil.rmtree, image_path)
-                        del self.image_data[image_id]
-                        log.info(f'Deleted image from cache with ID {image_id}')
+                    async with image_data.lock:
```

Comment:
This makes me think that we're not shielding enough. In the good case, this should be redundant because the `writer_lock` means that no coroutines are currently pulling or extracting any filesystems (since `localize_fs` grabs `image_lock.reader_lock`). I feel like we should instead fully shield `localize_fs` and be certain that locks pertaining to pulling/extracting filesystems are properly accounted for during cancellation. What do you think?